### PR TITLE
auto infer shape and dtype for onnx nodes

### PR DESCRIPTION
**Pull request by lucienwang1009**
1. Infer shapes and dtypes for ONNX nodes in the method `make_node` when creating a ONNX node. Only support the node whose inputs' shape and dtype are already set. Nodes involved in `body_graph`, such as Loop, Scan, If, are not supported for now. 
2. Explicitly set `parent_garph` if generated body graph is a sub-graph. 
3. Use `g.make_node` rather than `Node(helper.make_node())` in `insert_new_node_on_output`/`insert_new_node_on_input`.
4. Remove some useless `set_dtype` and `set_shape`.

### CODE REVIEW

**Comment by nbcsm** [comment]
add more background:
we found several dtype/shape missing issues when converting real models.
this pr tries to address some of such issues automatically.
so developer does not need to specify dtype/shape **everywhere**, only need to do it when necessary.

File: tf2onnx/graph.py [hunk 1]
```diff
@@ -252,6 +257,22 @@ def update_proto(self):
         if attr:
             self._op.attribute.extend(attr)
 
+    def build_onnx_node(self):
```
Comment (nbcsm) [inline tf2onnx/graph.py]: seems it is only used in infer_onnx_shape_dtype?
move to utils or a inline function of infer_onnx_shape_dtype?

File: tf2onnx/graph.py [hunk 2]
```diff
@@ -410,8 +431,49 @@ def make_node(self, op_type, inputs, attr=None, output_count=1, outputs=None, sk
...
+            return
+        log.debug("Infer shape and dtype for [%s]", node.name)
+        input_shapes = [self.get_shape(i) for i in node.input]
+        input_dtypes = [self.get_dtype(i) for i in node.input]
+
+        dtypes = {}
+        shapes = {}
+        try:
+            shapes, dtypes = utils.infer_onnx_shape_dtype(node, input_shapes, input_dtypes, self._opset)
+        except Exception:
```
[the remark this answers is not in the document]
Comment (lucienwang1009) [reply tf2onnx/graph.py]: Now, we won't throw any exception in `infer_onnx_shape_dtype`, so the exception caught here should be caused by ONNX infer_shape.

File: tf2onnx/graph.py [hunk 3]
```diff
@@ -410,8 +431,49 @@ def make_node(self, op_type, inputs, attr=None, output_count=1, outputs=None, sk
...
+            dtype = dtypes[output]
+            shape = shapes[output]
+            if dtype == TensorProto.UNDEFINED:
+                log.debug("Inferred dtype for [%s, type: %s] is None", node.name, node.type)
+            else:
+                existing_dtype = self.get_dtype(output)
+                if existing_dtype is not None and not override:
+                    dtype = existing_dtype
+                self.set_dtype(output, dtype)
+                log.debug("Set dtype of [%s] to %s", output, dtype)
```
[the remark this answers is not in the document]
Comment (lucienwang1009) [reply tf2onnx/graph.py]: class methods have only one new line inserted.

File: tf2onnx/utils.py [hunk 4]
```diff
@@ -347,3 +347,39 @@ def tf_name_scope(name):
...
     return [-1 for i in enumerate(shape)]
+
+
+def infer_onnx_shape_dtype(node, input_shapes, input_dtypes, opset=8):
+    """infer shapes and dtypes for outputs of the node"""
+    shapes = {}
+    dtypes = {}
+    inputs = []
+    outputs = []
+    for inp, shape, dtype in zip(node.input, input_shapes, input_dtypes):
```
[the remark this answers is not in the document]
Comment (lucienwang1009) [reply tf2onnx/utils.py]: Yes, you're right. ONNX infer_shape can work with None shape and `UNDEFINED` dtype as inputs.
I remove the requirement that shape and dtype of inputs cannot be None, and set unknown dtype of `tensor_info` to `UNDEFINED`.

**Comment by pengwa** [comment]
as we talked, I will look at once the set_nodes() is complete (if everybody agrees on that PR)

File: tf2onnx/utils.py [hunk 5]
```diff
@@ -437,3 +448,52 @@ def is_onnx_domain(domain):
...
+            copied_sub_graph = copy.deepcopy(sub_graph)
+            graph_proto = copied_sub_graph.make_graph("graph for " + node.name + " " + attr_name)
+            attr.append(helper.make_attribute(attr_name, graph_proto))
+    attr.extend([a for a in node.attr_onnx.values()])
+    if attr:
+        onnx_node.attribute.extend(attr)
+    return onnx_node
+
+
+def infer_onnx_shape_dtype(node, input_shapes, input_dtypes, opset=constants.PREFERRED_OPSET):
```
Comment (nbcsm) [inline tf2onnx/utils.py]: let's make opset required.

File: tf2onnx/utils.py [hunk 6]
```diff
@@ -437,3 +448,52 @@ def is_onnx_domain(domain):
     if domain is None or domain == "":
         return True
     return False
+
+
+def build_onnx_op(node):
```
Comment (nbcsm) [inline tf2onnx/utils.py]: make it an inline function of infer_onnx_shape_dtype

File: tf2onnx/graph.py [hunk 7]
```diff
@@ -535,6 +541,50 @@ def reset_nodes(self, ops):
         self._dtypes = remained_dtypes
         self._output_shapes = remained_shapes
 
+    def update_node_shape_dtype(self, node, override=True):
+        """try the best to infer shapes and dtypes for outputs of the node"""
```
Comment (nbcsm) [inline tf2onnx/graph.py]: skip if node is not in onnx domain.

File: tf2onnx/utils.py (hunk 5 above)
Comment (nbcsm) [reply tf2onnx/utils.py]: i am considering to move this to tf2onnx/schemas.py?
put these onnx def/schema related api calls together.

File: tf2onnx/utils.py [hunk 8]
```diff
@@ -437,3 +448,52 @@ def is_onnx_domain(domain):
...
+    """Build onnx op"""
+    onnx_node = helper.make_node(node.type, node.input, node.output, name=node.name)
+    # deal with attributes
+    attr = []
+    attr_graphs = node.get_body_graphs()
+    if attr_graphs:
```
Comment (nbcsm) [inline tf2onnx/utils.py]: i am curious about whether we can infer node with sub graph correctly.
you may want to add tests, e.g. a simple If/Loop/Scan?

File: tf2onnx/graph.py [hunk 9]
```diff
@@ -535,6 +541,50 @@ def reset_nodes(self, ops):
...
+
+        for output in node.output:
+            dtype = dtypes[output]
+            shape = shapes[output]
+            if dtype == TensorProto.UNDEFINED:
+                log.debug("Inferred dtype for [%s, type: %s] is None", node.name, node.type)
+            else:
+                existing_dtype = self.get_dtype(output)
+                if existing_dtype is not None:
+                    if override:
```
[the remark this answers is not in the document]
Comment (nbcsm) [reply tf2onnx/graph.py]: warning or just debug?
we already have too many warnings...

[the remark this answers is not in the document]
Comment (lucienwang1009) [reply tf2onnx/graph.py]: no worries. only see few in unittests.

File: tf2onnx/graph.py [hunk 10]
```diff
@@ -535,6 +541,50 @@ def reset_nodes(self, ops):
         self._dtypes = remained_dtypes
         self._output_shapes = remained_shapes
 
+    def update_node_shape_dtype(self, node, override=True):
```
[the remark this answers is not in the document]
Comment (lucienwang1009) [reply tf2onnx/graph.py]: em.. currently not use this option.

**Comment by lucienwang1009** [comment]
To make shape inference work, when creating new nodes, one needs to:
1. make sure the node's inputs are already in the graph.
2. if the node refers to outputs in parent graph, make sure the parent graph is set by `g.parent_graph = parent_graph`.

**Comment by lucienwang1009** [comment]
> please also add tests

added some tests stolen from ONNX: https://github.com/onnx/onnx/blob/master/onnx/test/shape_inference_test.py. :)

File: tests/test_shape_inference.py [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+            self.logger.error(ex)
+            raised = True
+        self.assertFalse(raised)
+
+        raised = False
+        for actual, inferred in zip(results, outputs):
+            actual_shape = actual.shape
+            inferred_shape = tuple(graph.get_shape(inferred))
+            try:
+                utils.merge_shapes(actual_shape, inferred_shape)
```
Comment (nbcsm) [inline tests/test_shape_inference.py]: are_shapes_compatible? so no need to try/catch

File: tests/test_shape_inference.py [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+    """
+
+    def _run_test_case(self, graph, feed_dict):
+        """Run model with onnxruntime and compare results' shape with internal shape inference."""
+        outputs = graph.outputs
+        results = []
+        raised = False
+        try:
+            results = self.run_backend(graph, outputs, feed_dict)
+        except Exception as ex:  # pylint: disable=broad-except
```
Comment (nbcsm) [inline tests/test_shape_inference.py]: no need to try/catch, just let it fail?

File: tf2onnx/graph.py [hunk 13]
```diff
@@ -419,7 +421,7 @@ def make_const(self, name, np_val, skip_conversion=False, raw=True):
         return node
 
     def make_node(self, op_type, inputs, attr=None, output_count=1, outputs=None, skip_conversion=True,
-                  op_name_scope=None, name=None, shapes=None, dtypes=None, domain=None):
+                  op_name_scope=None, name=None, shapes=None, dtypes=None, domain=None, auto_infer_shape_dtype=True):
```
Comment (nbcsm) [inline tf2onnx/graph.py]: auto_infer_shape_dtype => infer_shape_dtype?

File: tf2onnx/graph.py [hunk 14]
```diff
@@ -533,6 +537,68 @@ def reset_nodes(self, ops):
...
+        if node.domain != constants.ONNX_DOMAIN:
+            return
+
+        logger.debug("Infer shape and dtype for [%s]", node.name)
+        # NOTE: shape inference for some ops need the input values of the op, e.g., Reshape
+        # op needs the "Shape" value to infer output shape.
+        initializers = []
+        for i, inp in enumerate(node.inputs):
+            if not inp:
+                if utils.is_debug_mode():
```
Comment (nbcsm) [inline tf2onnx/graph.py]: logger.isEnabledFor(xxx)?

Comment (lucienwang1009) [reply tf2onnx/graph.py]: which level we should use? VERBOSE?

**Comment by nbcsm** [comment]
@lucienwang1009, i will merge it later after the random uniform bug is fixed.

**Pull request merged**